### PR TITLE
Add exclude of directory "db_analyze" to generate template to gpexpand

### PR DESCRIPTION
**Pull request by red1452**
If run `analyze_db` for any database, than for this database folder "db_analyze" is created. It may contain large volume of data. If after it create new segment using `gpexpand`, than folder "db_analyze" will be copied to new segment. This folder is useless for new segment. Because of this exclude "db_analyze" is added to `PgBaseBackup`.

Steps to reproduce the problem:
1) Create demo cluster.
2) Run `analyzedb -d postgres -a`.
3) Create config file "expand.conf" for `gpexpand`:
```
sdw1|sdw1|36101|<PATH_TO_GPDB_SRC>/gpAux/gpdemo/datadirs/dbfast4|9|3|p
sdw1|sdw1|36102|<PATH_TO_GPDB_SRC>/gpAux/gpdemo/datadirs/dbfast_mirror4|10|3|m
```
4) Run `gpexpand -i expand.conf`.
5) Done.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by RekGRpth — DISMISSED** [review]

**Comment by Stolb27** [comment]
1. There is already exising logic to exclude specific paths from backup [here](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/gpMgmt/bin/gppylib/commands/pg.py#L231-L245). @red1452, can you explain, why you prefer implementing this logic separately?
2. Also, I remembered one more [place ](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/filemap.c#L39-L90)with path exclusion. Maybe, extend gpexpand list with heaviest from them?

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature [hunk 1]
```diff
@@ -588,3 +588,22 @@ Feature: expand the cluster by adding more segments
         When the user runs gpexpand with a static inputfile for a single-node cluster with mirrors without ret code check
         Then gpexpand should return a return code of 0
         And gpexpand should print "One or more segments are either down or not in preferred role." to stdout
+
+    @gpexpand_mirrors
+    @gpexpand_segment
+    @gpexpand_verify_catalogs
```
Comment (Stolb27) [inline gpMgmt/test/behave/mgmt_utils/gpexpand.feature:594]: Can you explain meaning of these tags?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 2]
```diff
@@ -2016,6 +2016,30 @@ def impl(context, filename, contain, output):
...
+            curs = dbconn.execSQL(conn, "SELECT hostname, datadir FROM gp_segment_configuration WHERE content > -1;")
+            result = curs.fetchall()
+            segment_info = [(result[s][0], result[s][1]) for s in range(len(result))]
+    except Exception as e:
+        raise Exception("Could not retrieve segment information: %s" % e.message)
+
+    for info in segment_info:
+        host, datadir = info
+        filepath = os.path.join(datadir, filename)
+        cmd_str = 'ssh %s "ls -la %s | grep %s | wc -l"' % (host, datadir, filename)
```
Comment (Stolb27) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Class Command already has logic to run commands remotely.

**Comment by red1452** [comment]
> 2. Also, I remembered one more [place ](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/filemap.c#L39-L90)with path exclusion. Maybe, extend gpexpand list with heaviest from them?

This should not work. This directories are [set](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/filemap.c#L908-L915) at `filemap`. For directories, which are must to remove, system `rmdir` [is executed](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/file_ops.c#L256) from [remove_target](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/file_ops.c#L134) from [copy_executeFileMap](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/copy_fetch.c#L256) from [executeFileMap](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/fetch.c#L44) from [pg_rewind:main](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/pg_rewind.c#L373). `rmdir` deletes a directory, only if it is empty. But on practice, if add directory "db_analyze" to this place, I get empty directory "db_analyze" at new segment.

**Comment by red1452** [comment]
> 1. There is already exising logic to exclude specific paths from backup [here](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/gpMgmt/bin/gppylib/commands/pg.py#L231-L245). @red1452, can you explain, why you prefer implementing this logic separately?

It looks good and works good. I will change my solution by using this case.

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature (hunk 1 above)
Comment (red1452) [reply gpMgmt/test/behave/mgmt_utils/gpexpand.feature:594]: I think, these tags are used for filtering some test or group of tests.

**Comment by Stolb27** [comment]
> 2\. Also, I remembered one more [place ](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/bin/pg_rewind/filemap.c#L39-L90)with path exclusion. Maybe, extend gpexpand list with heaviest from them?

I've noticed that gpexpand only rename pg_log directory after restore without cleaning it up. I knew that gpexpand doesn't have problems with copying logs from master to new segments. I run pg_basebackup and confirmed that logs aren't copied:

```
 pg_basebackup -c fast -D /tmp/template -h localhost -p 6000 --xlog --target-gp-dbid=666 --progress --verbose
```

So, I suggested, that these directories are already filtered somewhere and found out appropriate code part for backend side of in the [basebackup.c](https://github.com/arenadata/gpdb/blob/8c9596cafd852a786c477bfe253321ba8f549398/src/backend/replication/basebackup.c#L189-L215). So filtering out on the gpexpand side is redundant for pg_log and backups directories.

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 3]
```diff
@@ -2016,6 +2016,33 @@ def impl(context, filename, contain, output):
...
+            curs = dbconn.execSQL(conn, "SELECT hostname, datadir FROM gp_segment_configuration WHERE content > -1;")
+            result = curs.fetchall()
+            segment_info = [(result[s][0], result[s][1]) for s in range(len(result))]
+    except Exception as e:
+        raise Exception("Could not retrieve segment information: %s" % e.message)
+
+    for info in segment_info:
+        host, datadir = info
+        filepath = os.path.join(datadir, filename)
+        cmd_str = 'ls -la %s | grep %s | wc -l' % (datadir, filename)
```
Comment (Stolb27) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: This approach may lead to false positive results:

```
givi@samsung-ivanov:~$ ls -la /tmp/2 | grep foo | wc -l
1
givi@samsung-ivanov:~$ ls /tmp/2
foobar
```

I think that check exact path is better:

```
givi@samsung-ivanov:~$ ls -la /tmp/2/foo
ls: cannot access '/tmp/2/foo': No such file or directory
```

Comment (RekGRpth) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: bash check directory exists
```sh
test -d /etc && echo true || echo false
true
```
```sh
test -d /etc1 && echo true || echo false
false
```

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 4]
```diff
@@ -2016,6 +2016,33 @@ def impl(context, filename, contain, output):
...
+            curs = dbconn.execSQL(conn, "SELECT hostname, datadir FROM gp_segment_configuration WHERE content > -1;")
+            result = curs.fetchall()
+            segment_info = [(result[s][0], result[s][1]) for s in range(len(result))]
+    except Exception as e:
+        raise Exception("Could not retrieve segment information: %s" % e.message)
+
+    for info in segment_info:
+        host, datadir = info
+        filepath = os.path.join(datadir, filename)
+        cmd_str = 'test -d %s && echo 1 || echo 0' % (filepath)
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: You should quote path like
```python
cmd_str = 'test -d "%s" && echo 1 || echo 0' % (filepath)
```

**Comment by red1452** [comment]
New commit message:
```
Add exclude of directory "db_analyze" to generate template to gpexpand (#476)

If run `analyze_db` for any database, than folder "db_analyze" is created.
It may contain large volume of data.
If after it create new segment using `gpexpand`, than folder "db_analyze"
will be copied to new segment. This folder is useless for new segment.
Because of this exclude "db_analyze" is added to `PgBaseBackup`.

Also new behave test on gpexpand was added to check that gpexpand does not
copy extra directories (all excludes from `PgBaseBackup`).
```

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 4 above)
Comment (RekGRpth) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: more, may be simpler to check exit code of test command?

**Review by Stolb27 — APPROVED** [review]

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**